sklearn/datasets/base.py

```diff
@@ -823,3 +825,134 @@ def _pkl_filepath(*args, **kwargs):
         basename += py3_suffix
     new_args = args[:-1] + (basename + ext,)
     return join(*new_args)
+
+
+class PartialURLOpener(urllib.FancyURLopener):
+    """A class to override urllib.FancyURLopener and
+    ignore HTTP error 206 (partial file being sent), since
+    that is what we expect when we resume the download
+    of a partial file
+    """
+
+    def http_error_206(self, url, fp, errcode, errmsg, headers, data=None):
+        """
+        Override HTTP Error 206 (partial file being sent). This error
+        indicates that the Range header is supported
+        """
+        # Ignore the expected "error" code
+        pass
+
+
+def _sha256(path):
+    """Calculate the sha256 hash of the file at path.
+
+    Parameters
+    -----------
+    path: String
+        Path of file to calculate SHA256 hash of.
+
+    Returns
+    -------
+    sha256hash : String
+        SHA256 hash of the file at the provided path.
+
+    """
+
+    sha256hash = hashlib.sha256()
+    chunk_size = 8192
+    with open(path, "rb") as f:
+        while 1:
+            buffer = f.read(chunk_size)
+            if not buffer:
+                break
+            sha256hash.update(buffer)
+    return sha256hash.hexdigest()
+
+
+def _validate_file_sha256(expected_checksum, path):
+    """Compare the SHA256 checksum of a file at a path with
+    an expected SHA256 checksum. If they do not match,
+    remove the file at path and throw a IOError.
+
+    Parameters
+    -----------
+    expected_checksum: String
+        Expected SHA256 checksum of file at path.
+
+    path: String
+        Path of file to compare SHA256 hash of.
+
+    """
+
+    if expected_checksum != _sha256(path):
+        # remove the corrupted file
+        remove(path)
+        raise IOError("{} has an SHA256 hash differing "
```

Comment:
Say in the error message that we deleted it? I'm on the fence about deleting it, but if we do, we should say so.